timemachine/cpp/src/wrap_kernels.cpp

```diff
@@ -381,6 +381,173 @@ void declare_potential(py::module &m) {
+
+                auto result = py::make_tuple(py::none(), py::none(), py::none(), py::none());
+                if (compute_du_dx) {
+                    py::array_t<double, py::array::c_style> py_du_dx(
+                        {coord_batches, param_batches, lambda_batches, N, D});
+                    for (unsigned int i = 0; i < du_dx.size(); i++) {
+                        py_du_dx.mutable_data()[i] = FIXED_TO_FLOAT<double>(du_dx[i]);
+                    }
+                    result[0] = py_du_dx;
+                }
+
+                if (compute_du_dp) {
+                    std::vector<ssize_t> pshape(params.shape(), params.shape() + params.ndim());
+                    // Remove the first dimension of the parameters shape to be consistent in ordering of return values
+                    pshape.erase(pshape.begin());
+                    // Append the new dimensions for the du_dps
+                    unsigned long int shape[] = {coord_batches, param_batches, lambda_batches};
+                    pshape.insert(pshape.begin(), shape, shape + 3);
+
+                    py::array_t<double, py::array::c_style> py_du_dp(pshape);
+                    for (unsigned int i = 0; i < K; i++) {
+                        pot.du_dp_fixed_to_float(N, P, &du_dp[0] + (i * P), py_du_dp.mutable_data() + (i * P));
+                    }
+                    result[1] = py_du_dp;
+                }
+
+                if (compute_du_dl) {
+                    py::array_t<double, py::array::c_style> py_du_dl({coord_batches, param_batches, lambda_batches});
+                    for (unsigned int i = 0; i < K; i++) {
+                        unsigned long long du_dl_sum = std::accumulate(
+                            &du_dl[0] + (i * N), &du_dl[0] + ((i + 1) * N), decltype(du_dl)::value_type(0));
+                        py_du_dl.mutable_data()[i] = FIXED_TO_FLOAT<double>(du_dl_sum);
+                    }
+                    result[2] = py_du_dl;
+                }
+
+                if (compute_u) {
+                    py::array_t<double, py::array::c_style> py_u({coord_batches, param_batches, lambda_batches});
+                    for (unsigned int i = 0; i < K; i++) {
+                        unsigned long long u_sum =
+                            std::accumulate(&u[0] + (i * N), &u[0] + ((i + 1) * N), decltype(u)::value_type(0));
+                        py_u.mutable_data()[i] = FIXED_TO_FLOAT<double>(u_sum);
+                    }
+                    result[3] = py_u;
+                }
+
+                return result;
+            },
+            py::arg("coords"),
+            py::arg("params"),
+            py::arg("boxes"),
+            py::arg("lambs"),
+            py::arg("compute_du_dx"),
+            py::arg("compute_du_dp"),
+            py::arg("compute_du_dl"),
+            py::arg("compute_u"),
+            R"pbdoc(
+        Execute the potential over a batch of coords, parameters and lambdas.
```

Comment:
Documentation: Would emphasize earlier that this function does `coords_batches * param_batches * lambda_batches` work.

(The return type documentation clarifies this, but a possible user error could still be to assume that it does `batch_size` work, with `coords_batches == param_batches == lambda_batches == batch_size`.)